Audio/SoundStreamInstance.cpp

```diff
@@ -0,0 +1,705 @@
+class SoundStreamInstance::Impl : public IVoiceNotify
+{
+public:
+    Impl(_In_ AudioEngine* engine,
+        WaveBank* waveBank,
+        uint32_t index,
+        SOUND_EFFECT_INSTANCE_FLAGS flags) noexcept(false) :
+        mBase(),
+        mWaveBank(waveBank),
+        mIndex(index),
+        mPlaying(false),
+        mLooped(false),
+        mEndStream(false),
+        mPrefetch(false),
+        mPackets{},
+        mCurrentDiskReadBuffer(0),
+        mCurrentPlayBuffer(0),
+        mStitchBytes(0),
+        mCurrentPosition(0),
+        mOffsetBytes(0),
+        mLengthInBytes(0),
+        mPacketSize(0),
+        mTotalSize(0)
+    {
+        assert(engine != nullptr);
+        engine->RegisterNotify(this, true);
+
+        char buff[64] = {};
+        auto wfx = reinterpret_cast<WAVEFORMATEX*>(buff);
+        assert(mWaveBank != nullptr);
+        mBase.Initialize(engine, mWaveBank->GetFormat(index, wfx, sizeof(buff)), flags);
+
+        WaveBankReader::Metadata metadata;
+        (void)mWaveBank->GetPrivateData(index, &metadata, sizeof(metadata));
+
+        mOffsetBytes = metadata.offsetBytes;
+        mLengthInBytes = metadata.lengthBytes;
+
+        mBufferEnd.reset(CreateEventEx(nullptr, nullptr, 0, EVENT_MODIFY_STATE | SYNCHRONIZE));
+        mBufferRead.reset(CreateEventEx(nullptr, nullptr, 0, EVENT_MODIFY_STATE | SYNCHRONIZE));
+        if (!mBufferEnd || !mBufferRead)
+        {
+            throw std::exception("CreateEvent");
+        }
+
+        ThrowIfFailed(AllocateStreamingBuffers(wfx));
+
+#ifdef VERBOSE_TRACE
+        DebugTrace("INFO (Streaming): packet size %zu, play length %zu\n", mPacketSize, mLengthInBytes);
+#endif
+
+        mPrefetch = true;
+        ThrowIfFailed(ReadBuffers());
+    }
+
+    virtual ~Impl() override
+    {
+        mBase.DestroyVoice();
+
+        if (mWaveBank && mWaveBank->GetAsyncHandle())
+        {
+            for (size_t j = 0; j < MAX_BUFFER_COUNT; ++j)
+            {
+                (void)CancelIoEx(mWaveBank->GetAsyncHandle(), &mPackets[j].request);
+            }
+        }
+
+        if (mBase.engine)
+        {
+            mBase.engine->UnregisterNotify(this, false, true);
+            mBase.engine = nullptr;
+        }
+
+        memset(mPackets, 0, sizeof(mPackets));
+        mPacketSize = 0;
+    }
+
+    Impl(Impl&&) = default;
+    Impl& operator= (Impl&&) = default;
+
+    Impl(Impl const&) = delete;
+    Impl& operator= (Impl const&) = delete;
+
+    void Play(bool loop)
+    {
+        if (!mBase.voice)
+        {
+            if (!mWaveBank)
+                return;
+
+            char buff[64] = {};
+            auto wfx = reinterpret_cast<WAVEFORMATEX*>(buff);
+            mBase.AllocateVoice(mWaveBank->GetFormat(mIndex, wfx, sizeof(buff)));
+        }
+
+        if (!mBase.Play())
+            return;
+
+        mLooped = loop;
+        mEndStream = false;
+
+        if (!mPrefetch)
+        {
+            mCurrentPosition = 0;
```

Comment:
On construction we prefetch the first few packets to give low-latency start times when you call ``Play``. Typically in *DirectX Tool Kit for Audio* if you call ``Play`` a second time it's supposed to restart at the beginning, so this little ``mPrefetch`` switch lets that work while not breaking prefetch.